mlflow/store/tracking/file_store.py

```diff
@@ -1319,3 +1326,372 @@ def _read_helper(root, file_name, attempts_remaining=2):
+                trace_path = os.path.join(traces_path, request_id)
+                # Do not throw if the trace doesn't exist
+                if exists(trace_path):
+                    shutil.rmtree(trace_path)
+                    deleted_traces += 1
+            return deleted_traces
+
+    def search_traces(
+        self,
+        experiment_ids: List[str],
+        filter_string: Optional[str] = None,
+        max_results: int = SEARCH_TRACES_DEFAULT_MAX_RESULTS,
+        order_by: Optional[List[str]] = None,
+        page_token: Optional[str] = None,
+    ):
+        """
+        Return traces that match the given list of search expressions within the experiments.
+
+        Args:
+            experiment_ids: List of experiment ids to scope the search.
+            filter_string: A search filter string. Supported filter keys are `name`,
+                           `status` and `timestamp_ms`.
+            max_results: Maximum number of traces desired.
+            order_by: List of order_by clauses. Supported sort key is `timestamp_ms`. By default
+                      we sort by timestamp_ms DESC.
+            page_token: Token specifying the next page of results. It should be obtained from
+                a ``search_traces`` call.
+
+        Returns:
+            A tuple of a list of :py:class:`TraceInfo <mlflow.entities.TraceInfo>` objects that
+            satisfy the search expressions and a pagination token for the next page of results.
+            If the underlying tracking store supports pagination, the token for the
+            next page may be obtained via the ``token`` attribute of the returned object; however,
+            some store implementations may not support pagination and thus the returned token would
+            not be meaningful in such cases.
+        """
+        if max_results > SEARCH_MAX_RESULTS_THRESHOLD:
+            raise MlflowException(
+                "Invalid value for request parameter max_results. It must be at "
+                f"most {SEARCH_MAX_RESULTS_THRESHOLD}, but got value {max_results}",
+                INVALID_PARAMETER_VALUE,
+            )
+        traces = []
+        for experiment_id in experiment_ids:
+            trace_infos = self._list_trace_infos(experiment_id)
+            traces.extend(trace_infos)
+        filtered = SearchTracesUtils.filter(traces, filter_string)
+        sorted_traces = SearchTracesUtils.sort(filtered, order_by)
+        traces, next_page_token = SearchTracesUtils.paginate(sorted_traces, page_token, max_results)
+        return traces, next_page_token
+
+    def _list_trace_infos(self, experiment_id):
+        experiment_path = self._get_experiment_path(experiment_id, assert_exists=True)
+        traces_path = os.path.join(experiment_path, FileStore.TRACES_FOLDER_NAME)
+        trace_paths = list_all(traces_path, lambda x: os.path.isdir(x), full_path=True)
+        trace_infos = []
+        for trace_path in trace_paths:
```

Comment:
Is there any possibility this cause memory crush? Trace info is not too heavy object but might be a bit dangerous if we have millions of traces. Do you think it's safer to add some guardrail for how many traces to keep in memory (or how many traces can be saved within experiment)?

Reply:
That's a good point, then we should add some check to calculate the existing traces number when starting trace. While traces should be similar to runs, and we haven't received any issues complaining about search_runs, so I guess we're fine for now?